test-harness/src/test/java/com/netflix/conductor/tests/integration/End2EndGrpcTests.java

```diff
@@ -190,40 +190,117 @@ public void testAll() throws Exception {
+
+        WorkflowDef workflowDefinition = new WorkflowDef();
+        workflowDefinition.setName("testEphemeralWorkflow");
+
+        WorkflowTask workflowTask1 = createWorkflowTask("storedTaskDef1");
+        WorkflowTask workflowTask2 = createWorkflowTask("storedTaskDef2");
+
+        workflowDefinition.getTasks().add(workflowTask1);
+        workflowDefinition.getTasks().add(workflowTask2);
+
+        String workflowExecutionName = "ephemeralWorkflow";
+        StartWorkflowRequest workflowRequest = new StartWorkflowRequest()
+                .withName(workflowExecutionName)
+                .withWorkflowDef(workflowDefinition);
+
+        String workflowId = workflowClient.startWorkflow(workflowRequest);
+        assertNotNull(workflowId);
+
+        Workflow workflow = workflowClient.getWorkflow(workflowId, true);
+        WorkflowDef ephemeralWorkflow = workflow.getWorkflowDefinition();
+        assertNotNull(ephemeralWorkflow);
+        assertEquals(workflowDefinition.getName(), ephemeralWorkflow.getName());
+    }
+
+    @Test
+    public void testEphemeralWorkflowsWithEphemeralTasks() throws Exception {
+        WorkflowDef workflowDefinition = new WorkflowDef();
+        workflowDefinition.setName("testEphemeralWorkflowWithEphemeralTasks");
+
+        WorkflowTask workflowTask1 = createWorkflowTask("ephemeralTask1");
+        TaskDef taskDefinition1 = createTaskDefinition("ephemeralTaskDef1");
+        workflowTask1.setTaskDefinition(taskDefinition1);
+
+        WorkflowTask workflowTask2 = createWorkflowTask("ephemeralTask2");
+        TaskDef taskDefinition2 = createTaskDefinition("ephemeralTaskDef2");
+        workflowTask2.setTaskDefinition(taskDefinition2);
+
+        workflowDefinition.getTasks().add(workflowTask1);
+        workflowDefinition.getTasks().add(workflowTask2);
+
+        String workflowExecutionName = "ephemeralWorkflowWithEphemeralTasks";
+        StartWorkflowRequest workflowRequest = new StartWorkflowRequest()
+                .withName(workflowExecutionName)
+                .withWorkflowDef(workflowDefinition);
+
+        String workflowId = workflowClient.startWorkflow(workflowRequest);
+        assertNotNull(workflowId);
+
+        Workflow workflow = workflowClient.getWorkflow(workflowId, true);
+        WorkflowDef ephemeralWorkflow = workflow.getWorkflowDefinition();
+        assertNotNull(ephemeralWorkflow);
+        assertEquals(workflowDefinition.getName(), ephemeralWorkflow.getName());
+
+        List<WorkflowTask> ephemeralTasks = ephemeralWorkflow.getTasks();
+        assertEquals(2, ephemeralTasks.size());
+        for (WorkflowTask ephemeralTask : ephemeralTasks) {
+            assertNotNull(ephemeralTask.getTaskDefinition());
+        }
+
+    }
+
+    // Helper method for creating task definitions on the server
+    private List<TaskDef> createAndRegisterTaskDefinitions(String prefixTaskDefinition, int numberOfTaskDefinitions) {
```

Comment:
**Suggestion:** Perhaps this can be `protected` and moved to the parent/abstract class?  If there is logic specific to the client type then you could implement that as an abstract method?

Reply:
This is a tricky one:
Client types are different and they are the implementations themselves. It would be great if they both implemented an interface but it is not the case. Also, I believe these are generated meaning I cannot change them.
One thing i thought yesterday to abstract this was to use Consumer<...> but will play around with it today.
Will update this thread with any ideas

Reply:
I see what you are saying now. Sure, moving it up 👍